opnode/p2p/gossip.go

```diff
@@ -0,0 +1,403 @@
+package p2p
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/binary"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	lru "github.com/hashicorp/golang-lru"
+
+	"github.com/ethereum-optimism/optimistic-specs/opnode/l2"
+	"github.com/ethereum-optimism/optimistic-specs/opnode/rollup"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/golang/snappy"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	pb "github.com/libp2p/go-libp2p-pubsub/pb"
+)
+
+func init() {
+	// TODO: a PR is open to make this configurable upstream as option instead of having to override a global
+	// https://github.com/libp2p/go-libp2p-pubsub/pull/484
+	pubsub.TimeCacheDuration = 80 * pubsub.GossipSubHeartbeatInterval
+}
+
+const maxGossipSize = 1 << 20
+const maxOutboundQueue = 256
+const maxValidateQueue = 256
+const globalValidateThrottle = 512
+
+// Message domains, the msg id function uncompresses to keep data monomorphic,
+// but invalid compressed data will need a unique different id.
+
+var MessageDomainInvalidSnappy = [4]byte{0, 0, 0, 0}
+var MessageDomainValidSnappy = [4]byte{1, 0, 0, 0}
+
+const MaxGossipSize = 1 << 20
+
+func blocksTopicV1(cfg *rollup.Config) string {
+	return fmt.Sprintf("/optimism/%s/0/blocks", cfg.L2ChainID.String())
+}
+
+// BuildSubscriptionFilter builds a simple subscription filter,
+// to help protect against peers spamming useless subscriptions.
+func BuildSubscriptionFilter(cfg *rollup.Config) pubsub.SubscriptionFilter {
+	return pubsub.NewAllowlistSubscriptionFilter(blocksTopicV1(cfg)) // add more topics here in the future, if any.
+}
+
+var msgBufPool = sync.Pool{New: func() any {
+	// note: the topic validator concurrency is limited, so pool won't blow up, even with large pre-allocation.
+	x := make([]byte, 0, MaxGossipSize)
+	return &x
+}}
+
+// BuildMsgIdFn builds a generic message ID function for gossipsub that can handle compressed payloads,
+// mirroring the eth2 p2p gossip spec.
+func BuildMsgIdFn(cfg *rollup.Config) pubsub.MsgIdFunction {
+	return func(pmsg *pb.Message) string {
+		valid := false
+		var data []byte
+		// If it's a valid compressed snappy data, then hash the uncompressed contents.
+		// The validator can throw away the message later when recognized as invalid,
+		// and the unique hash helps detect duplicates.
+		dLen, err := snappy.DecodedLen(pmsg.Data)
```

Comment:
Is this immune to zip-bomb like attacks?

Reply:
See the `dLen <= MaxGossipSize` below. It detects the length before actually allocating or uncompressing it. If the length is too large it will handle it like a message with invalid compression: the data stays as-is, and the message ID is computed over the data with a different domain (to avoid collisions with valid compressed data). The validation pipeline will later take care of the message being too large / compressed wrong.